src/dev/lupio/lupio_rtc.cc

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2021 The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dev/lupio/lupio_rtc.hh"
+
+#include "base/time.hh"
+#include "debug/LupioRTC.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioRTC.hh"
+
+namespace gem5
+{
+
+using namespace RiscvISA;
+
+LupioRTC::LupioRTC(const Params &params) :
+    BasicPioDevice(params, params.pio_size),
+    time(params.time)
+{
+    gettimeofday(&tv, NULL);
+    gmtime_r(&tv.tv_sec, &time);
+
+    DPRINTF(LupioRTC, "Real-time clock set to %s", asctime(&time));
```

Comment:
`tv` and `time` haven't been declared, and I'm still unclear on what this constructor is trying to do. Can you explain?

Reply:
Currently, we have both `tv` and `time` defined in `lupio_rtc.hh`, because we use them in both the constructor and in LupioRTCRead.  In the constructor, we just wanted to grab the current time so we could use a debug flag to look at the time being saved while we were testing, since it was the only function that we were sure was being called while we still weren't finished.  However, we could now instead move this print statement to LupioRTCRead, make `tv` and `time` local variables to that function, and empty the constructor if that makes more sense to do?

Reply:
Unless you really initialize the time in the constructor (in which case the device model needs to be somewhat rewritten to take that in account --but in that case, your variable should be part of the class itself), then you could just empty the constructor and the printf could be move to the read function (or even dropped altogether).